Framework/Algorithms/src/MultipleScatteringCorrection.cpp

```diff
@@ -0,0 +1,403 @@
+  // size_t distGraber.m_numVolumeElements : The number of volume elements
+  size_t numVolumeElements = distGraber.m_numVolumeElements;
+
+  // L2D needs to be calculated w.r.t the detector
+  // L12 is independent from the detector, therefore can be cached outside
+  // - L12 is a upper off-diagonal matrix
+  // NOTE: if the sample size/volume is too large, we might need to use openMP
+  //       to parallelize the calculation
+  size_t len_l12 = numVolumeElements * (numVolumeElements - 1) / 2;
+  std::vector<double> sample_L12s(len_l12, 0.0);
+  for (size_t i = 0; i < numVolumeElements; ++i) {
+    for (size_t j = i + 1; j < numVolumeElements; ++j) {
+      const V3D dist = distGraber.m_elementPositions[i] - distGraber.m_elementPositions[j];
+      size_t idx = calcLinearIdxFromUpperTriangular(numVolumeElements, i, j);
+      sample_L12s[idx] = dist.norm();
+    }
+  }
+
+  // debug L12 matrix
+  // NOTE: this will come in handy when we start considering container and sample interaction,
+  //       do not remove it.
+  // std::ostringstream msg;
+  // msg << "\n";
+  // for (size_t i = 0; i < numVolumeElements; ++i) {
+  //   for (size_t j = 0; j < numVolumeElements; ++j) {
+  //     if (i < j) {
+  //       size_t idx = numVolumeElements * (numVolumeElements - 1) / 2 -
+  //                    (numVolumeElements - i) * (numVolumeElements - i - 1) / 2 + j - i - 1;
+  //       msg << sample_L12s[idx] << " ";
+  //     } else {
+  //       msg << "x ";
+  //     }
+  //   }
+  //   msg << '\n';
+  // }
+  // g_log.notice(msg.str());
+
+  const auto &spectrumInfo = m_inputWS->spectrumInfo();
+  const auto numHists = static_cast<int64_t>(m_inputWS->getNumberHistograms());
+  const auto specSize = static_cast<int64_t>(m_inputWS->blocksize());
+  Progress prog(this, 0.0, 1.0, numHists);
+  // -- loop over the spectra/detectors
+  PARALLEL_FOR_IF(Kernel::threadSafe(*m_inputWS, *m_outputWS))
+  for (int64_t i = 0; i < numHists; ++i) {
+    PARALLEL_START_INTERUPT_REGION
+
+    // copy bins from input workspace to output workspace
+    m_outputWS->setSharedX(i, m_inputWS->sharedX(i));
+
+    // locate the spectrum and its detector
+    if (!spectrumInfo.hasDetectors(i)) {
+      g_log.information() << "Spectrum " << i << " does not have a detector defined for it\n";
+      continue;
+    }
+    const auto &det = spectrumInfo.detector(i);
+
+    // compute L2D
+    std::vector<double> sample_L2Ds(numVolumeElements, 0.0);
+    calculateL2Ds(distGraber, det, sample_L2Ds);
+
+    const auto wavelengths = m_inputWS->points(i);
+    // these need to have the minus sign applied still
+    const auto sampleLinearCoefAbs = m_material.linearAbsorpCoef(wavelengths.cbegin(), wavelengths.cend());
+    // TODO:
+    // - implement the multiple scattering correction for heterogenous media (container)
+
+    auto &output = m_outputWS->mutableY(i);
+    // -- loop over the wavelength points every m_xStep
+    for (int64_t j = 0; j < specSize; j += m_xStep) {
```

Comment:
Please rename these variables to highlight that they are for the wavelength bins.

Reply:
inner loop index has been refactored as suggested